pkg/schedule/affinity/policy.go

```diff
@@ -0,0 +1,232 @@
+		case store.IsDisconnected():
+			unavailableStores[store.GetID()] = storeDisconnected
+		case store.IsLowSpace(lowSpaceRatio):
+			unavailableStores[store.GetID()] = storeLowSpace
+		case store.IsPreparing():
+			unavailableStores[store.GetID()] = storePreparing
+		}
+		// Note: We intentionally do NOT check:
+		// - IsSlow(): Performance issue, not availability issue
+	}
+	return unavailableStores
+}
+
+func (m *Manager) getGroupAvailabilityChanges(unavailableStores map[uint64]storeCondition) (isUnavailableStoresChanged bool, groupAvailabilityChanges map[string]groupAvailability) {
+	var logEntries []logEntry
+	groupAvailabilityChanges = make(map[string]groupAvailability)
+
+	// Validate whether unavailableStores has changed.
+	m.RLock()
+	isUnavailableStoresChanged = !maps.Equal(unavailableStores, m.unavailableStores)
+	if !isUnavailableStoresChanged {
+		m.RUnlock()
+		return false, nil
+	}
+
+	// Analyze which Groups have changed availability
+	// Collect log messages to print after releasing lock
+	for _, groupInfo := range m.groups {
+		var unavailableStore uint64
+		var maxCondition storeCondition
+		for _, storeID := range groupInfo.VoterStoreIDs {
+			if condition, ok := unavailableStores[storeID]; ok && (!condition.affectsLeaderOnly() || storeID == groupInfo.LeaderStoreID) {
+				if unavailableStore == 0 || condition > maxCondition {
+					unavailableStore = storeID
+					maxCondition = condition
+				}
+			}
+		}
+		newAvailability := maxCondition.groupAvailability()
+		if newAvailability != groupInfo.GetAvailability() {
+			groupAvailabilityChanges[groupInfo.ID] = newAvailability
+			if unavailableStore != 0 {
+				logEntries = append(logEntries, logEntry{
+					level:            zapcore.WarnLevel,
+					groupID:          groupInfo.ID,
+					unavailableStore: unavailableStore,
+					availability:     newAvailability,
+				})
+			} else {
+				logEntries = append(logEntries, logEntry{
+					level:   zapcore.InfoLevel,
+					groupID: groupInfo.ID,
+				})
+			}
+		}
+	}
+	m.RUnlock()
+
+	// Log after releasing lock
+	for _, entry := range logEntries {
+		switch entry.level {
+		case zapcore.WarnLevel:
+			log.Warn("affinity group invalidated due to unavailable stores",
+				zap.String("group-id", entry.groupID),
+				zap.Uint64("unavailable-store", entry.unavailableStore),
+				zap.String("availability", entry.availability.String()))
+		case zapcore.InfoLevel:
+			log.Info("affinity group become available", zap.String("group-id", entry.groupID))
+		}
+	}
+
+	return
+}
+
+func (m *Manager) setGroupAvailabilityChanges(unavailableStores map[uint64]storeCondition, groupAvailabilityChanges map[string]groupAvailability) {
+	m.Lock()
+	defer m.Unlock()
+	m.unavailableStores = unavailableStores
+	for groupID, availability := range groupAvailabilityChanges {
+		m.updateGroupAvailabilityLocked(groupID, availability)
+	}
+}
+
+func (m *Manager) hasUnavailableStore(leaderStoreID uint64, voterStoreIDs []uint64) error {
```

Comment:
In general, the function named hasxxxx should return the store ID rather than an error.